fea/initial_release

mimir-d opened the pull request:
- this PR is mainly cleanup and documentation before making the first stable release version
- it's easier to go commit by commit when reading

mimir-d committed:
Add option to set manual timezone

- there are a number of places where string formatted timestamps are
  used; these contain timezone info as per spec
- add new config option to set this timezone, instead of using the local
  one by default
- add tests to cover and example

Signed-off-by: mimir-d <mimir-d@users.noreply.github.com>

mimir-d committed:
Add docstrings for all user facing methods

- docstrings specify parameters, types and reference back to spec text

Signed-off-by: mimir-d <mimir-d@users.noreply.github.com>

mimir-d committed:
Add installation and usage instructions

- add sample code to readme, along with pip installation

Signed-off-by: mimir-d <mimir-d@users.noreply.github.com>

mimir-d committed:
Add api reference docs; generator process

- fix reST docstrings so that sphinx generates the api ref nicely
- add sphinx with markdown generator to write api ref docs
- add reference to readme to point to the api ref index
- add the sphinx generated markdown files so that theyre usable from gh

Signed-off-by: mimir-d <mimir-d@users.noreply.github.com>

mimir-d committed:
Split examples/demo.py into multiple modules

- this makes it easier to read the examples in order of complexity
- also useful later on when checking against the schema validator

Signed-off-by: mimir-d <mimir-d@users.noreply.github.com>

mimir-d committed:
Add missing config doc; samples output

- and other misc cleanup

Signed-off-by: mimir-d <mimir-d@users.noreply.github.com>

njordr commented on README.md:
````diff
@@ -1,28 +1,168 @@
+For general usage, the following steps should be sufficient to get the latest stable version:
+
+```bash
+# [option 1] install system-wide, from any shell
````
I would put a different block code for each option, to avoid confusion for those (like me) go and start immediately to type commands

njordr commented on README.md:
```diff
@@ -1,28 +1,168 @@
+$ cd ocp-diag-python
+$ git checkout dev # dev branch has the latest code
+
+# [option 1] install to system, inside the ocp-diag-python dir
```
as my previous comment

njordr commented on README.md:
```diff
@@ -1,28 +1,168 @@
+**Minimum python version is currently py37.**
+
+For general usage, the following steps should be sufficient to get the latest stable version:
+
```
could we add an option with poetry? it's widely used at the moment

njordr commented on README.md:
````diff
@@ -1,28 +1,168 @@
+To setup any of these aspects:
+```py
+import ocptv.output as tv
+
````
The import for FileWriter is missing

mimir-d replied:
ok, so it wasnt obvious that `FileWriter` is usercode. i'll replace it with a `StdoutWriter`. though i could also argue that these should come with the lib, incl rolling file; i'll leave this as a feature request

dalwise commented on ocptv/output/config.py:
```diff
@@ -68,22 +70,44 @@ def enable_runtime_checks(self, enable: bool):
     :param Writer writer: if provided, set the output channel writer.
     :param bool enable_runtime_checks: if provided, enables or disables runtime type checks.
+    :param tzinfo timezone: if provided, sets the timezone for the output formatted datetime fields.
+        To reset to local timezone, `None` value should be used.
```
Spec specifies "The only guidance provided is that the implementation shall conform to this format, and should be referenced to UTC time when possible.".

Shouldn't the default therefore be UTC? Local timezone is non-deterministic in areas with summer time during the hour before and after the time gets moved back.

mimir-d replied:
moved everything to UTC; you are on point about the spec

dalwise commented on ocptv/output/config.py:
```diff
@@ -93,6 +117,9 @@ def config(
     if enable_runtime_checks is not None:
         _config.enable_runtime_checks = enable_runtime_checks
 
+    if timezone is not _NOT_SET:
```
If `ty.TYPE_CHECKING` is `True` then `_NOT_SET` would be `None` and one would not be able to reset the timezone by using `None` as the docstring states, right?

mimir-d replied:
`ty.TYPE_CHECKING` is true only when... type checking (as in, when it's being parsed by `mypy`, in this case). So everything in that branch of the if doesnt affect runtime behavior. I had to introduce that `_NOT_SET` because `None` is a valid value for the parameter, which makes `ty.Optional` unusable

mimir-d commented:
note: im waiting for all people to have a look, then i'll address all of the feedback

mimir-d committed:
Add installation instructions for poetry

- split installation steps into separate bash codeblocks in README
- add minimal pyproject.toml config for poetry to recognize this as a
  project (necessary for install from git repo)

Signed-off-by: mimir-d <mimir-d@users.noreply.github.com>

mimir-d committed:
Review iteration

- fix default timezone to be UTC as per spec
- remove ref to unknown type FileWriter in README

Signed-off-by: mimir-d <mimir-d@users.noreply.github.com>
